<a href="https://colab.research.google.com/github/Lottus0412/Estrutura_de_Dados_II/blob/main/Aula_3_Estruturas_Dinamicas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Conteúdo de Revisão — Aula 03: Ponteiros, Referências e Estruturas Dinâmicas em Python
Estrutura de Dados
Duração: 1h30
Linguagem: Python
Tema: Ponteiros de memória, referências e estruturas dinâmicas.



Exercício 1 — Referências

In [8]:

a = [10, 20, 30]
b = b


b.append(40)

print("a:", a)
print("b:", b)
print("id(a):", id(a))
print("id(b):", id(b))

a: [10, 20, 30]
b: [10, 20, 30, 40, 40]
id(a): 132972846748608
id(b): 132973548457152


**Qual será a saída?**

a: [10, 20, 30, 40]

b: [10, 20, 30, 40]

id(a): 132973548456512

id(b): 132973548456512

**a e b representam o mesmo objeto? Por quê?**
A e B representam o mesmo objeto, pois agregam o mesmo resultado.

**O que aconteceria com b = a.copy()?**

a: [10, 20, 30]

b: [10, 20, 30, 40]

id(a): 132972848004864

id(b): 132973548457152


Exercício 2 — Construindo uma cadeia



In [9]:
n1 = print("A")
n2 = print("B")
n3 = print("C")

A
B
C


 Exercício 3 — Depuração

In [16]:
class Node:
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None


n1 = Node(10)
n2 = Node(20)
n3 = Node(30)

n1.proximo = n2
n2.proximo = n3

atual = n1

while atual is not None:
    print(atual.valor)
    atual = atual.proximo

10
20
30


**Qual é o problema?**
O problema é que falta atualizar a variável atual dentro do while. Assim, ela continua apontando para o mesmo nó.

**Por que o programa não termina?**
Porque atual continua sendo n1, então atual nunca chega a None. O while fica repetindo para sempre.

**Qual linha deve ser acrescentada?**
Dentro do while, deve ser acrescentada: atual = atual.proximo

**Qual será a saída depois da correção?**
10
20
30

Problema real — Sistema de atendimento de uma clínica

In [17]:
class Paciente:
    def __init__(self, nome, idade, prioridade):
        self.nome = nome
        self.idade = idade
        self.prioridade = prioridade


class Node:
    def __init__(self, paciente):
        self.paciente = paciente
        self.proximo = None


class FilaAtendimento:
    def __init__(self):
        self.inicio = None
        self.fim = None
        self.quantidade = 0

    # 1. Adicionar paciente
    def adicionar(self, paciente):
        novo = Node(paciente)

        if self.fim is None:
            self.inicio = novo
            self.fim = novo
        else:
            self.fim.proximo = novo
            self.fim = novo

        self.quantidade += 1

    # 2. Listar pacientes
    def listar(self):
        atual = self.inicio

        while atual is not None:
            paciente = atual.paciente
            print(paciente.nome, "-", paciente.idade, "anos -", paciente.prioridade)
            atual = atual.proximo

    # 3. Atender o primeiro paciente
    def atender(self):
        if self.inicio is None:
            return None

        paciente = self.inicio.paciente
        self.inicio = self.inicio.proximo

        if self.inicio is None:
            self.fim = None

        self.quantidade -= 1

        return paciente

    # 4. Verificar se está vazia
    def vazia(self):
        return self.inicio is None

    # 5. Informar quantidade
    def tamanho(self):
        return self.quantidade


fila = FilaAtendimento()

fila.adicionar(Paciente("Ana", 32, "Normal"))
fila.adicionar(Paciente("Bruno", 70, "Prioridade"))
fila.adicionar(Paciente("Carlos", 45, "Normal"))

fila.listar()

paciente = fila.atender()

print("Atendendo:", paciente.nome)
print("Quantidade restante:", fila.tamanho())

Ana - 32 anos - Normal
Bruno - 70 anos - Prioridade
Carlos - 45 anos - Normal
Atendendo: Ana
Quantidade restante: 2
